src/DynamoCore/Graph/Nodes/CustomNodes/Function.cs

```diff
@@ -514,5 +572,69 @@ protected override bool UpdateValueCore(UpdateValueParams updateValueParams)
 
             return base.UpdateValueCore(updateValueParams);
         }
+
+        private bool TryParseOutputExpression(string expression, out IdentifierNode outputIdentifier, out string comment)
+        {
+            outputIdentifier = null;
+            comment = null;
+
+            var resolver = workspaceElementResolver ?? ElementResolver;
+            var parseParam = new ParseParam(GUID, expression + ";", resolver);
+
+            if (EngineController.CompilationServices.PreCompileCodeBlock(ref parseParam) 
+                && parseParam.ParsedNodes.Any())
+            {
+                var parsedComments = parseParam.ParsedComments;
+                if (parsedComments.Any())
+                {
+                    comment = String.Join("\n", parsedComments.Select(c => (c as CommentNode).Value));
+                }
+
+                if (parseParam.ParsedNodes.Count() > 1)
+                {
+                    this.Warning(Properties.Resources.WarningInvalidOutput);
+                }
+
+                var node = parseParam.ParsedNodes.First() as BinaryExpressionNode;
+                if (node == null)
+                {
+                    if (parseParam.Errors.Any())
+                    {
+                        this.Error(Properties.Resources.WarningInvalidOutput);
+                    }
+                }
+                else
+                {
+                    var leftIdent = node.LeftNode as IdentifierNode;
+                    var rightIdent = node.RightNode as IdentifierNode;
+
+                    // "x" will be compiled to "temp_guid = x;"
+                    if (leftIdent != null && leftIdent.Value.StartsWith(Constants.kTempVarForNonAssignment))
+                    {
+                        outputIdentifier = rightIdent;
+                    }
+                    // "x:int" will be compiled to "x:int = tTypedIdent0;"
+                    else if (rightIdent != null && rightIdent.Value.StartsWith(Constants.kTempVarForTypedIdentifier))
+                    {
+                        outputIdentifier = leftIdent;
+                    }
+                    else
+                    {
+                        if (parseParam.Errors.Any())
+                        {
+                            this.Error(parseParam.Errors.First().Message);
+                        }
+                        else
+                        {
+                            this.Warning(Properties.Resources.WarningInvalidOutput);
+                        }
+                        outputIdentifier = leftIdent;
```

Comment:
if output is invalid, we still try to display whatever string in the output port... so technically the variable name 'outputIdentifier' is not quite correct...